src/xds/provider/agones.rs

```diff
@@ -0,0 +1,438 @@
+                        }),
+                        ..<_>::default()
+                    })),
+                    metadata: Some(Metadata {
+                        filter_metadata: HashMap::from([(
+                            String::from(QUILKIN_URL),
+                            Struct {
+                                fields: BTreeMap::from([(String::from(QUILKIN_TOKEN_KEY), tokens)]),
+                            },
+                        )]),
+                        ..<_>::default()
+                    }),
+                    ..<_>::default()
+                }
+            })
+            .collect();
+
+        Ok(Cluster {
+            name: DEFAULT_CLUSTER_NAME.into(),
+            load_assignment: Some(ClusterLoadAssignment {
+                cluster_name: DEFAULT_CLUSTER_NAME.into(),
+                endpoints: vec![LocalityLbEndpoints {
+                    lb_endpoints,
+                    ..<_>::default()
+                }],
+                ..<_>::default()
+            }),
+            cluster_discovery_type: Some(ClusterDiscoveryType::Type(DiscoveryType::Static as i32)),
+            ..<_>::default()
+        })
+    }
+}
+
+#[tonic::async_trait]
+impl crate::xds::DiscoveryServiceProvider for AgonesProvider {
+    async fn discovery_request(
+        &self,
+        _node_id: &str,
+        version: u64,
+        kind: ResourceType,
+        _names: &[String],
+    ) -> Result<DiscoveryResponse, tonic::Status> {
+        let resources = match kind {
+            ResourceType::Endpoint => {
+                let cluster = self.get_cluster().await?;
+                let mut buf = Vec::new();
+                buf.reserve(cluster.encoded_len());
+
+                cluster.encode(&mut buf).unwrap();
+
+                vec![Any {
+                    type_url: crate::xds::ENDPOINT_TYPE.into(),
+                    value: buf,
+                }]
+            }
+            ResourceType::Listener => {
+                let listener = self.get_listener().await?;
+                let mut buf = Vec::new();
+                buf.reserve(listener.encoded_len());
+
+                listener.encode(&mut buf).unwrap();
+
+                vec![Any {
+                    type_url: kind.type_url().into(),
+                    value: buf,
+                }]
+            }
+            kind => {
+                return Err(tonic::Status::internal(format!(
+                    "Quilkin currently does not support {} requests",
+                    kind.type_url()
+                )))
+            }
+        };
+
+        Ok(DiscoveryResponse {
+            version_info: version.to_string(),
+            resources,
+            type_url: kind.type_url().into(),
+            ..<_>::default()
+        })
+    }
+}
+
+#[derive(kube::CustomResource, Clone, Debug, Deserialize, Serialize, JsonSchema)]
+#[kube(
+    group = "agones.dev",
+    version = "v1",
+    kind = "GameServer",
+    status = "GameServerStatus",
```

Comment:
Technically we don't have an actual k8s/status value for `GameServer` (for our use case it reduces hits to the API server) - but since we're not updating a `GameServer`, it likely doesn't matter for our use case (I'm guessing it might be required for this build toolchain? 🤔)

Reply:
Are you sure, because there is a `GameServerStatus` type that is defined here. https://github.com/googleforgames/agones/blob/9fa432c24d5aefbfcd6cf7eef61fcd4b35034663/pkg/apis/agones/v1/gameserver.go#L221

Reply:
Yes, his is correct, we have a status value - but it's not registered a a `/status` subresource on `GameServer`, because reasons.

Details: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#subresources

(Basically it's a nice way of updating only a section of a CRD - but the way we treat `GameServers` in Agones, it's better if we update GameServer's entire value as a single transaction, so we don't do this for this resource)

If it's working for this, it doesn't actually matter - it might have been an issue if we were trying to update `/status` values on the `GameSever` as it might use the wrong API.

Mostly this is just a point of interest, and not a blocking issue in the PR.